runtime/convertValues.go

```diff
@@ -31,117 +31,145 @@ import (
 
+	results := exportResults{}
+
 	for i, field := range event.Fields {
-		fields[i] = exportValueWithInterpreter(field.Value, field.Interpreter())
+		fields[i] = exportValueWithInterpreter(field.Value, field.Interpreter(), results)
 	}
 
 	eventType := exportType(event.Type, map[sema.TypeID]cadence.Type{}).(*cadence.EventType)
 	return cadence.NewEvent(fields).WithType(eventType)
 }
 
-func exportValueWithInterpreter(value interpreter.Value, inter *interpreter.Interpreter) cadence.Value {
-	switch v := value.(type) {
-	case interpreter.VoidValue:
-		return cadence.NewVoid()
-	case interpreter.NilValue:
-		return cadence.NewOptional(nil)
-	case *interpreter.SomeValue:
-		return exportSomeValue(v, inter)
-	case interpreter.BoolValue:
-		return cadence.NewBool(bool(v))
-	case *interpreter.StringValue:
-		return cadence.NewString(v.Str)
-	case *interpreter.ArrayValue:
-		return exportArrayValue(v, inter)
-	case interpreter.IntValue:
-		return cadence.NewIntFromBig(v.ToBigInt())
-	case interpreter.Int8Value:
-		return cadence.NewInt8(int8(v))
-	case interpreter.Int16Value:
-		return cadence.NewInt16(int16(v))
-	case interpreter.Int32Value:
-		return cadence.NewInt32(int32(v))
-	case interpreter.Int64Value:
-		return cadence.NewInt64(int64(v))
-	case interpreter.Int128Value:
-		return cadence.NewInt128FromBig(v.ToBigInt())
-	case interpreter.Int256Value:
-		return cadence.NewInt256FromBig(v.ToBigInt())
-	case interpreter.UIntValue:
-		return cadence.NewUIntFromBig(v.ToBigInt())
-	case interpreter.UInt8Value:
-		return cadence.NewUInt8(uint8(v))
-	case interpreter.UInt16Value:
-		return cadence.NewUInt16(uint16(v))
-	case interpreter.UInt32Value:
-		return cadence.NewUInt32(uint32(v))
-	case interpreter.UInt64Value:
-		return cadence.NewUInt64(uint64(v))
-	case interpreter.UInt128Value:
-		return cadence.NewUInt128FromBig(v.ToBigInt())
-	case interpreter.UInt256Value:
-		return cadence.NewUInt256FromBig(v.ToBigInt())
-	case interpreter.Word8Value:
-		return cadence.NewWord8(uint8(v))
-	case interpreter.Word16Value:
-		return cadence.NewWord16(uint16(v))
-	case interpreter.Word32Value:
-		return cadence.NewWord32(uint32(v))
-	case interpreter.Word64Value:
-		return cadence.NewWord64(uint64(v))
-	case interpreter.Fix64Value:
-		return cadence.Fix64(v)
-	case interpreter.UFix64Value:
-		return cadence.UFix64(v)
-	case *interpreter.CompositeValue:
-		return exportCompositeValue(v, inter)
-	case *interpreter.DictionaryValue:
-		return exportDictionaryValue(v, inter)
-	case interpreter.AddressValue:
-		return cadence.NewAddress(v)
-	case *interpreter.StorageReferenceValue:
-		return exportStorageReferenceValue(v)
-	case interpreter.LinkValue:
-		return exportLinkValue(v, inter)
-	case interpreter.PathValue:
-		return exportPathValue(v)
-	case interpreter.TypeValue:
-		return exportTypeValue(v, inter)
-	case interpreter.CapabilityValue:
-		return exportCapabilityValue(v, inter)
+type exportResults map[interpreter.Value]cadence.Value
```

Comment:
What is this type for? I see it written to but not read from.

Reply:
The type is used for the "results" parameter of `exportValueWithInterpreter`, and read from at the top of the function call: https://github.com/onflow/cadence/pull/400/files#diff-0f63ad66a8e46ed831efe45b45a08d6aR59.

As the export is recursive, the results parameter prevents cycles.